src/Microsoft.AspNetCore.JsonPatch/Internal/ParsedPath.cs

```diff
@@ -12,14 +14,19 @@ public struct ParsedPath
 
         private readonly string[] _segments;
 
+        private static readonly Regex escapingRegex = new Regex("~(?<code>0|1)", RegexOptions.Compiled);
+
         public ParsedPath(string path)
         {
             if (path == null)
             {
                 throw new ArgumentNullException(nameof(path));
             }
 
-            _segments = path.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+            _segments = path
+                .Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries)
+                .Select(segment => escapingRegex.Replace(segment, (match) => match.Groups["code"].Value == "0" ? "~" : "/"))
```

Comment:
Hi, I've quickly wrote a [console app](https://github.com/sandorfr/random-stuff/blob/master/JsonPointer/ParseJsonPointerTest/src/ParseJsonPointerTest/Program.cs) to test two different approaches : 
Replace is definitely better than Regex. 

However, a custom implementation reducing memory allocation is slightly more performant than Split + replace. It also gives us a chance to catch bad escaping sequences.

```
RUN 1:
ParsePathWithSplitAndRegex :329 ms.
ParsePathWithSplitAndReplace :146 ms.
ParsePathWithCustomImplementation :133 ms.
RUN 2:
ParsePathWithSplitAndRegex :263 ms.
ParsePathWithSplitAndReplace :82 ms.
ParsePathWithCustomImplementation :84 ms.
RUN 3:
ParsePathWithSplitAndRegex :154 ms.
ParsePathWithSplitAndReplace :109 ms.
ParsePathWithCustomImplementation :119 ms.
RUN 4:
ParsePathWithSplitAndRegex :205 ms.
ParsePathWithSplitAndReplace :138 ms.
ParsePathWithCustomImplementation :97 ms.
RUN 5:
ParsePathWithSplitAndRegex :229 ms.
ParsePathWithSplitAndReplace :87 ms.
ParsePathWithCustomImplementation :69 ms.
```

Let met know what you think.